### PR TITLE
Change error message for ty param in const

### LINKED ISSUE #76701: min_const_generics: ty param in const incorrect help
```rust
#![feature(min_const_generics)]

fn test<T>() -> [u8; std::mem::size_of::<T>()] {
    todo!()
}
```
emits
```
error: generic parameters must not be used inside of non trivial constant values
 --> src/lib.rs:3:42
  |
3 | fn test<T>() -> [u8; std::mem::size_of::<T>()] {
  |                                          ^ non-trivial anonymous constants must not depend on the parameter `T`
  |
  = help: it is currently only allowed to use either `T` or `{ T }` as generic constants
```

The help is incorrect here, as type parameters in constants are not supported at all with `min_const_generics`.

The error is emitted here https://github.com/rust-lang/rust/blob/b5f55b7e1562286798358730e06b0ba13b4c0ea6/compiler/rustc_resolve/src/diagnostics.rs#L469-L485

We probably want to change `ParamInNonTrivialAnonConst` to instead contain an `Option` which is only `Some` for const parameters.

For type parameters we can instead emit a note saying
```
note: type parameters are currently not permitted in anonymous constants
```

If you need more help feel free to ask either here on on zulip.

### PR DESCRIPTION
**Pull request by hameerabbasi**
This PR introduces the following changes:

* Change error message for type param in a const expression when using
`min_const_generics`
* Change `ParamInNonTrivialAnonConst` to contain an extra `bool` used for
distinguishing whether the passed-in symbol is a type or a value.

Fixes #76701

### CODE REVIEW

File: compiler/rustc_resolve/src/diagnostics.rs [hunk 1]
```diff
@@ -479,10 +479,13 @@ impl<'a> Resolver<'a> {
                     ),
                 );
                 err.help(
-                    &format!("it is currently only allowed to use either `{0}` or `{{ {0} }}` as generic constants", name)
+                    &format!("note: type parameters are currently not permitted in anonymous constants")
                 );
                 err
             }
+            ResolutionError::ParamInNonTrivialAnonConst(None) => {
+                todo!()
```
Comment (hameerabbasi) [inline compiler/rustc_resolve/src/diagnostics.rs:487]: I'm unsure what to do here, or where it'll be needed.

File: compiler/rustc_resolve/src/diagnostics.rs [hunk 2]
```diff
@@ -479,10 +479,13 @@ impl<'a> Resolver<'a> {
                     ),
                 );
                 err.help(
-                    &format!("it is currently only allowed to use either `{0}` or `{{ {0} }}` as generic constants", name)
+                    &format!("note: type parameters are currently not permitted in anonymous constants")
                 );
```
Comment (jyn514) [inline compiler/rustc_resolve/src/diagnostics.rs:483]: Using both 'help' and 'note: ' is confusing, try `err.note("type parameters ...")` instead maybe.

File: compiler/rustc_resolve/src/diagnostics.rs (hunk 1 above)
Comment (jyn514) [reply compiler/rustc_resolve/src/diagnostics.rs:487]: I saw you only construct this with `Some` - why make it an `Option` then?

File: compiler/rustc_resolve/src/lib.rs [hunk 3]
```diff
@@ -2639,7 +2639,7 @@ impl<'a> Resolver<'a> {
                                         self.report_error(
                                             span,
                                             ResolutionError::ParamInNonTrivialAnonConst(
```
Comment (lcnr) [inline compiler/rustc_resolve/src/lib.rs:2641]: you are dealing with a type parameter here, so we want to use `None` here.

Comment (hameerabbasi) [reply compiler/rustc_resolve/src/lib.rs]: As discussed out-of-band, we needed the name of the symbol here, so we used an additional bool, and passed it in.

File: compiler/rustc_resolve/src/diagnostics.rs (hunk 1 above)
Comment (hameerabbasi) [reply compiler/rustc_resolve/src/diagnostics.rs]: We changed it to include a `bool`, which is now being used.

File: compiler/rustc_resolve/src/diagnostics.rs [hunk 4]
```diff
@@ -478,9 +478,19 @@ impl<'a> Resolver<'a> {
...
                     ),
                 );
-                err.help(
-                    &format!("it is currently only allowed to use either `{0}` or `{{ {0} }}` as generic constants", name)
-                );
+
+                if is_type {
+                    err.note(&format!(
+                        "type parameters are currently not permitted in anonymous constants"
+                    ));
```
Comment (lcnr) [inline compiler/rustc_resolve/src/diagnostics.rs:485]:
```suggestion
                    err.note("type parameters are currently not permitted in anonymous constants");
```

Comment (hameerabbasi) [reply compiler/rustc_resolve/src/diagnostics.rs:485]: I had it that way originally, I ran `rustfmt` and this popped out. Would you rather I changed it back?

Comment (lcnr) [reply compiler/rustc_resolve/src/diagnostics.rs:485]: It's not about changing the formatting here. The relevant part is the removal of `&format!`.

**Pull request merged**